soc/arm/nuvoton_npcx/npcx7/power.c

```diff
@@ -0,0 +1,157 @@
+ *
+ * This file contains the drivers of NPCX Power Manager Modules that improves
+ * the efficiency of ec operation by adjusting the chip’s power consumption to
+ * the level of activity required by the application. The following table
+ * summarizes the main properties of the various power states and shows the
+ * activity levels of the various clocks while in these power states.
+ *
+ * +--------------------------------------------------------------------------+
+ * | Power State | LFCLK | HFCLK | APB/AHB |  Core  | RAM/Regs   | VCC | VSBY |
+ * |--------------------------------------------------------------------------|
+ * | Active      | On    | On    | On      | Active | Active     | On  | On   |
+ * | Idle (wfi)  | On    | On    | On      | Wait   | Active     | On  | On   |
+ * | Sleep       | On    | On    | Stop    | Stop   | Preserved  | On  | On   |
+ * | Deep Sleep  | On    | Stop  | Stop    | Stop   | Power Down | On  | On   |
+ * | Stand-By    | Off   | Off   | Off     | Off    | Off        | Off | On   |
+ * +--------------------------------------------------------------------------+
+ *
+ * LFCLK - Low-Frequency Clock. Its frequency is fixed to 32kHz.
+ * HFCLK - High-Frequency (PLL) Clock. Its frequency is configured to OSC_CLK.
+ *
+ * Based on the follwoing criteria:
+ *
+ * - A delay of 'Instant' wake-up from 'Deep Sleep' is 20 us.
+ * - A delay of 'Standard' wake-up from 'Deep Sleep' is 3.43 ms.
+ * - Max residency time in Deep Sleep for 'Instant' wake-up is 200 ms
+ * - Min Residency time in Deep Sleep for 'Instant' wake-up is 61 us
+ * - The unit to determine power state residency policy is tick.
+ *
+ * this driver implements one power state, PM_STATE_SUSPEND_TO_IDLE, with
+ * two sub-states for power management system.
+ * Sub-state 0 - "Deep Sleep" mode with “Instant” wake-up if residency time
+ *               is greater or equal to 1 ms
+ * Sub-state 1 - "Deep Sleep" mode with "Standard" wake-up if residency time
+ *               is greater or equal to 201 ms
+ *
+ * INCLUDE FILES: soc_clock.h
+ */
+
+#include <zephyr.h>
+#include <power/power.h>
+#include <soc.h>
+
+#include <logging/log.h>
+LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
+
+/* The steps that npcx ec enters sleep/deep mode and leaves it. */
+#define NPCX_ENTER_SYSTEM_SLEEP() ({                                           \
+	__asm__ volatile (                                                     \
+	"push {r0-r5}\n"     /* Save the registers used for delay */           \
+	"wfi\n"              /* Enter sleep mode after receiving wfi */        \
+	"ldm %0, {r0-r5}\n"  /* Add a delay before instructions fetching */    \
+	"pop {r0-r5}\n"      /* Restore the registers used for delay */        \
+	"isb\n"              /* Flush the cpu pipelines */                     \
+	:: "r" (CONFIG_SRAM_BASE_ADDRESS)); /* A valid addr used for delay */  \
+	})
+
+#if defined(CONFIG_SOC_POWER_MANAGEMENT_TRACE)
+uint32_t cnt_sleep0;
+uint32_t cnt_sleep1;
+#endif
```

Comment:
If we just have these as static will they not be taken out by the optimizer anyway if `!IS_ENABLED(CONFIG_SOC_POWER_MANAGEMENT_TRACE)`? Putting these in will allow us to use the more readable `IS_ENABLED` further down.

Reply:
Good idea. I will replace them with IS_ENABLED() macro later.

Reply:
Done.